pkg/asset/upgrade/bundle.go

```diff
@@ -0,0 +1,489 @@
+	metadata := &Metadata{
+		Version: g.appCfg.Config.OcpRelease.Version,
+		Arch:    *g.appCfg.Config.OcpRelease.CpuArchitecture,
+		Release: releaseImageRef,
+		Images:  maps.Values(imageMap),
+	}
+	err = g.writeMetadata(ctx, metadata)
+	if err != nil {
+		return err
+	}
+
+	// Write the bundle:
+	g.logger.Infof("Writing bundle to '%s' ...", g.bundleFile())
+	err = g.writeBundle(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Write the digest:
+	g.logger.Infof("Writing digest to '%s' ...", g.digestFile())
+	err = g.writeDigest(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (g *bundleGenerator) startRegistry(ctx context.Context) (result *registry.RegistryServer,
+	err error) {
+	result, err = registry.NewRegistryServer().
+		SetLogger(logrusr.New(g.logger)).
+		SetListenAddr("localhost:0").
+		SetRootDir(g.tmpDir).
+		Build()
+	if err != nil {
+		return
+	}
+	err = result.Start(ctx)
+	return
+}
+
+// findImages finds the release metadata image and from that the rest of the release images. It
+// returns a map where the keys are the tags and the values are the full image references.
+func (g *bundleGenerator) findImages(ctx context.Context) (releaseImageRef string,
+	imageMap map[string]string, err error) {
+	// Calculate the release metadata image tag and reference:
+	releaseImageTag := fmt.Sprintf(
+		"%s-%s",
+		g.appCfg.Config.OcpRelease.Version,
+		*g.appCfg.Config.OcpRelease.CpuArchitecture,
+	)
+	releaseImageRef = fmt.Sprintf(
+		"%s/%s:%s",
+		bundleReleaseImageDomain, bundleReleaseImagePath, releaseImageTag,
+	)
+
+	// Use the 'oc adm release info' command to extract the release metadata JSON from the
+	// release metadata image:
+	ocPath, err := exec.LookPath("oc")
+	if err != nil {
+		return
+	}
+	ocStdout := &bytes.Buffer{}
+	ocStderr := &bytes.Buffer{}
+	ocCmd := exec.Cmd{
+		Path: ocPath,
+		Args: []string{
+			"oc",
+			"adm",
+			"release",
+			"info",
+			"--output=json",
+			releaseImageRef,
+		},
+		Stdout: ocStdout,
+		Stderr: ocStderr,
+	}
+	err = ocCmd.Run()
+	g.logger.WithFields(logrus.Fields{
+		"args":   ocCmd.Args,
+		"stdout": ocCmd.String(),
+		"stderr": ocCmd.String(),
+		"code":   ocCmd.ProcessState.ExitCode(),
+	}).Debug("Executed 'oc' command")
+	if err != nil {
+		return
+	}
+
+	// Extract from the release metadata JSON the digest of the release metadata image, so that
+	// we can later download it using the digest instead of the tag that we calculated earlier:
+	var releaseImageDigest string
+	err = g.jqTool.QueryBytes(
+		`.digest`,
+		ocStdout.Bytes(), &releaseImageDigest,
+	)
+	if err != nil {
+		return
+	}
+	releaseImageRef = fmt.Sprintf(
+		"%s/%s@%s",
+		bundleReleaseImageDomain, bundleReleaseImagePath, releaseImageDigest,
+	)
+
+	// Extract from the release metadata JSON the references of the rest of the images that are
+	// part of the release:
+	type ImageInfo struct {
+		Tag string `json:"tag"`
+		Ref string `json:"ref"`
+	}
+	var imageInfos []ImageInfo
+	err = g.jqTool.QueryBytes(
```

Comment:
Can we maybe use mapping.txt file that oc-mirror is creating: https://github.com/openshift-kni/telco-ran-tools/blob/main/cmd/download.go#L345C1-L345C1
Might save some work here...